core/dbt/contracts/connection.py

```diff
@@ -106,7 +106,9 @@ class Credentials(
     Replaceable,
     metaclass=abc.ABCMeta
 ):
-    database: str
+    # Most DBs have this as required, but BigQuery is Optional, and mypy
+    # doesn't seem to allow overriding the type in `BigQueryCredentials`
+    database: Optional[str]
```

Comment:
We also set `database` to `Optional[str]` [in dbt-spark](https://github.com/fishtown-analytics/dbt-spark/blob/bde721eda38bdccf3f7b5c1bd0ac11b270c644d4/dbt/adapters/spark/connections.py#L51), which inherits/replaces the base `Credentials` dataclass, too. I wonder why mypy doesn't mind it there

Reply:
Great point, thanks @jtcohen6 . I think I'm confusing errors — the error it issued was actually `Attributes without a default cannot follow attributes with one`. Let's try the latest push

Reply:
Also does repo actually run mypy? 
I see a lot of type annotations in the code, but also a lot of errors when running mypy? Mostly in the `test` path

Reply:
Yes, we run mypy, but only on the `core/dbt` path:
https://github.com/fishtown-analytics/dbt/blob/5ba5271da99bc1ef7fbad2f7d0b45634087300fa/tox.ini#L15

Reply:
https://github.com/fishtown-analytics/dbt/issues/2926